File: MELA/test/testME_madgraph.cpp
#include <iostream>
#include <fstream>
#include "Mela.h"
#include "TVar.hh"
#include "TCouplingsBase.hh"
#include "TFile.h"
#include "TTree.h"
#include "TLorentzVector.h"
#include "TTreeReader.h"
#include "TTreeReaderValue.h"
#include "TTreeReaderArray.h"
using namespace std;

TLorentzVector ptEtaPhiVector(double pt, double eta, double phi, double m){ //constructs pt, eta, phi TLorentzVector
    TLorentzVector vec = TLorentzVector();
    vec.SetPtEtaPhiM(pt, eta, phi, m);
    return vec;
}



int main(int argc, char const *argv[])
{
    // TLorentzVector g1 = TLorentzVector(0.0000000E+00,  0.0000000E+00,  0.5000000E+03, 0.5000000E+03);
    // TLorentzVector g2 = TLorentzVector(0.0000000E+00,  0.0000000E+00,  -0.5000000E+03, 0.5000000E+03);

    // SimpleParticleCollection_t* mothers = new SimpleParticleCollection_t();
    // mothers->push_back(SimpleParticle_t(21, g1));
    // mothers->push_back(SimpleParticle_t(21, g2));

    // TLorentzVector l1 = TLorentzVector(-0.2210069E+02,  0.4008035E+02, -0.7580543E+02, 0.8855133E+02);
    // TLorentzVector l2 = TLorentzVector(-0.1038496E+03, -0.3019338E+03,  0.7649492E+02, 0.3283294E+03);
    // TLorentzVector l3 = TLorentzVector(-0.1058810E+03, -0.9770964E+02,  0.4954839E+02, 0.1523581E+03);
    // TLorentzVector l4 = TLorentzVector( 0.2318313E+03,  0.3595630E+03, -0.5023788E+02, 0.4307611E+03);

    // SimpleParticleCollection_t* daughters = new SimpleParticleCollection_t();
    // daughters->push_back(SimpleParticle_t(11, l2));
    // daughters->push_back(SimpleParticle_t(-11, l1));
    // daughters->push_back(SimpleParticle_t(11, l4));
    // daughters->push_back(SimpleParticle_t(-11, l3));

    TFile* dataFile = TFile::Open("/eos/home-m/msrivast/CMSSW_14_0_0/src/JHUGenMELA/MELA/test/PS_HIGGS_MADGRAPH.root");
    // TFile* dataFile = TFile::Open("/eos/home-m/msrivast/CMSSW_14_0_0/src/HexUtils/SimulationTools/lhe_tools/HIGGS_MAD.root");


    TTreeReader myReader("tree", dataFile);

    TTreeReaderArray<float> LHEDaughterId(myReader,   "LHEDaughterId");
    TTreeReaderArray<float> LHEDaughterPt(myReader,   "LHEDaughterPt");
    TTreeReaderArray<float> LHEDaughterEta(myReader,  "LHEDaughterEta");
    TTreeReaderArray<float> LHEDaughterPhi(myReader,  "LHEDaughterPhi");
    TTreeReaderArray<float> LHEDaughterMass(myReader, "LHEDaughterMass");
    
    TTreeReaderArray<float> LHEAssociatedParticleId(myReader,   "LHEAssociatedParticleId");
    TTreeReaderArray<float> LHEAssociatedParticlePt(myReader,   "LHEAssociatedParticlePt");
    TTreeReaderArray<float> LHEAssociatedParticleEta(myReader,  "LHEAssociatedParticleEta");
    TTreeReaderArray<float> LHEAssociatedParticlePhi(myReader,  "LHEAssociatedParticlePhi");
    TTreeReaderArray<float> LHEAssociatedParticleMass(myReader, "LHEAssociatedParticleMass");
    
    TTreeReaderArray<float> LHEMotherId(myReader, "LHEMotherId");
    TTreeReaderArray<float> LHEMotherPx(myReader, "LHEMotherPx");
    TTreeReaderArray<float> LHEMotherPy(myReader,  "LHEMotherPy");
    TTreeReaderArray<float> LHEMotherPz(myReader,  "LHEMotherPz");
    TTreeReaderArray<float> LHEMotherE(myReader,  "LHEMotherE");


    TTreeReaderValue<float> M4L(myReader, "M4L");

    Mela m = Mela(13, 125, TVar::SILENT);


    vector<vector<float>> madprobs;
    madprobs.push_back(vector<float>());
    madprobs.push_back(vector<float>());

    vector<vector<float>> jhugenprobs;
    jhugenprobs.push_back(vector<float>());
    jhugenprobs.push_back(vector<float>());
    vector<float> m4l;

    int counter = 0;
    int NEVENT = 10000;
    while(myReader.Next() && counter < NEVENT){
        if((counter % 100) == 0){
            cerr << counter << endl;
        }
        SimpleParticleCollection_t* mother_collection = new SimpleParticleCollection_t();
        SimpleParticleCollection_t* daughter_collection = new SimpleParticleCollection_t();
        SimpleParticleCollection_t* associated_collection = new SimpleParticleCollection_t();

        int i = 0;
        for(i = 0; i < 2; i++){
            mother_collection->push_back(
                SimpleParticle_t((int)(LHEMotherId[i]), TLorentzVector(LHEMotherPx[i], LHEMotherPy[i], LHEMotherPz[i], LHEMotherE[i]))
            );
        }

        for(i = 0; i < 4; i++){
            daughter_collection->push_back(
                SimpleParticle_t((int)(LHEDaughterId[i]),ptEtaPhiVector(LHEDaughterPt[i], LHEDaughterEta[i], LHEDaughterPhi[i], LHEDaughterMass[i]))
            );
        }

        for(int mat_el : {1, 3}){
            for(int setup : {0,1}){
                m.setProcess(TVar::SelfDefine_spin0, static_cast<TVar::MatrixElement>(mat_el), TVar::ZZGG);
                m.setInputEvent(daughter_collection, 0, mother_collection, true, (bool)(mat_el == 3));

                if(mat_el == 1){
                    if(setup == 0){
                        m.selfDHggcoupl[0][gHIGGS_GG_2][0] = 1;
                        m.selfDHzzcoupl[0][gHIGGS_VV_1][0] = 2;
                        m.selfDHzzcoupl[0][gHIGGS_VV_4][0] = 0;
                    } else if(setup == 1){
                        m.selfDHggcoupl[0][gHIGGS_GG_2][0] = 1;
                        m.selfDHzzcoupl[0][gHIGGS_VV_4][0] = 1;
                        m.selfDHzzcoupl[0][gHIGGS_VV_1][0] = 0;
                    }
                } else if(mat_el == 3){
                    if(setup == 1){
                        // From Lexicon
                        m.mdl_chwtil = -6.34078;
                        m.mdl_chbtil = -1.90674;
                        m.mdl_chg = -8.24752;
                        m.mdl_chwbtil = -6.9542;
                        m.mdl_chbox = -16.495;
                    }
                }
                float ans = 0;
                m.computeP(ans);
                if(mat_el == 1){
                    jhugenprobs[setup].push_back(ans);
                } else if(mat_el == 3){
                    madprobs[setup].push_back(ans);
                }
                m.resetInputEvent();
            }
        }
        m4l.push_back(*M4L);
        counter++;
        delete mother_collection;
        delete daughter_collection;
        delete associated_collection;
    }

    dataFile->Close();
    ofstream outputFile;
    cout << "WRITTEN SUCCESSFULLY" << endl;
    cout << m4l.size() << " EVENTS" << endl;
    outputFile.open("probs_output.csv");
    outputFile << "jhugenprob_g1, jhugenprob_g4, madprob_g1, madprob_g4, m4l\n";
    for(int j = 0; j < m4l.size(); j++){
        outputFile << jhugenprobs[0][j] << ", " << jhugenprobs[1][j] << ", " << madprobs[0][j] << ", " << madprobs[1][j] << ", " << m4l[j] << "\n";
    }
    outputFile.close();

    return 0;
}

File: MELA/test/testME_madgraph.py
import Mela
import uproot
import numpy
import tqdm
# particles = [
#         [0.5000000E+03,  0.0000000E+00,  0.0000000E+00,  0.5000000E+03 ],
#         [0.5000000E+03,  0.0000000E+00,  0.0000000E+00, -0.5000000E+03 ],
#         [0.8855133E+02, -0.2210069E+02,  0.4008035E+02, -0.7580543E+02 ],
#         [0.3283294E+03, -0.1038496E+03, -0.3019338E+03,  0.7649492E+02 ],
#         [0.1523581E+03, -0.1058810E+03, -0.9770964E+02,  0.4954839E+02 ],
#         [0.4307611E+03,  0.2318313E+03,  0.3595630E+03, -0.5023788E+02 ]
#     ]
# particles = [
#         [5.37086667033E+02, 0.00000000000E+00,  0.00000000000E+00,  5.37086667033E+02  ],
#         [7.27293062837E+00, 0.00000000000E+00,  0.00000000000E+00, -7.27293062837E+00  ],
#         [2.11966451221E+01,  1.78742924427E+01, -1.12102312034E+01, -2.03424438331E+00 ],
#         [7.02133017289E+01,  3.86248305415E+01, -3.12040320381E+01, -4.96421050079E+01 ],
#         [9.29989707133E+00,  5.37669110106E+00, -4.15860490296E-01, -7.57669704638E+00 ],
#         [1.23934598003E+02, -4.84989558626E+01, -4.60240644895E+01, -1.04352390245E+02 ]
#     ]

# ids = [21, 21, -11, 11, -11, 11]

m = Mela.Mela(13, 125, Mela.VerbosityLevel.SILENT)
m.setProcess(Mela.Process.SelfDefine_spin0, Mela.MatrixElement.MADGRAPH, Mela.Production.ZZGG)

dataFile = uproot.open("SM_HIGGS_JHUGEN.root")
tree = dataFile["tree"]

NEVENT = tree.num_entries

data = tree.arrays(
    [
        "LHEDaughterId", "LHEDaughterPt", "LHEDaughterEta", "LHEDaughterPhi", "LHEDaughterMass", 
        "LHEAssociatedParticleId", "LHEAssociatedParticlePt", "LHEAssociatedParticleEta", "LHEAssociatedParticlePhi", "LHEAssociatedParticleMass",
        "LHEMotherId", "LHEMotherPz", "LHEMotherE", "M4L", "Phid", "costheta1d", "costheta2d", "Phi1d"
    ],
    library='np', entry_stop=NEVENT + 1
)

f = open("probs_output.csv", "w+")
probs = [
    [[], []], #jhugen
    [[], []]  #madgraph
]

jhugen_couplings = [
    {
        "ghg2":[1,0],
        "ghz1":[1,0]
    },
    {
        "ghg2":[1,0],
        "ghz4":[1,0]
    }
]

madgraph_couplings = [
    {
        
    },
    {
        "mdl_chbtil":1,
        "mdl_chwtil":1,
        "mdl_chwbtil":1
    }
]
for n, (jhucoupl, madcoupl) in enumerate(zip(jhugen_couplings, madgraph_couplings)):
    for mat_el in [Mela.MatrixElement.JHUGen, Mela.MatrixElement.MADGRAPH]:
        if mat_el.value == 1:
            coupl = jhucoupl
        elif mat_el.value == 3:
            coupl = madcoupl

        for i in tqdm.tqdm(range(NEVENT), total=NEVENT):   
            mothers = Mela.SimpleParticleCollection_t(
                data["LHEMotherId"][i], 
                [0]*len(data["LHEMotherId"][i]), 
                [0]*len(data["LHEMotherId"][i]), 
                data["LHEMotherPz"][i], 
                data["LHEMotherE"][i], 
                False
            )
            
            daughters = Mela.SimpleParticleCollection_t(
                data["LHEDaughterId"][i], 
                data["LHEDaughterPt"][i], 
                data["LHEDaughterEta"][i], 
                data["LHEDaughterPhi"][i], 
                data["LHEDaughterMass"][i], 
                True
            )
            
            # associated = Mela.SimpleParticleCollection_t(
            #     data["LHEAssociatedParticleId"][i], 
            #     data["LHEAssociatedParticlePt"][i], 
            #     data["LHEAssociatedParticleEta"][i], 
            #     data["LHEAssociatedParticlePhi"][i], 
            #     data["LHEAssociatedParticleMass"][i], 
            #     True
            # )
            
            associated = Mela.SimpleParticleCollection_t()
            
            m.setProcess(Mela.Process.SelfDefine_spin0, mat_el, Mela.Production.ZZGG)
            m.setInputEvent(daughters, associated, mothers, True, (mat_el.value == 3))
            
            for coupling, value in coupl.items():
                setattr(m, coupling, value)
            
            if mat_el.value == 1:
                probs[0][n].append(m.computeP(False))
            elif mat_el.value == 3:
                probs[1][n].append(m.computeP(False))
            else:
                raise ValueError("ERROR!")
            m.resetInputEvent()

f.write("jhugenprob_g1, jhugenprob_g4, madprob_g1, madprob_g4, m4l\n")

# for n, (j,m) in enumerate(zip(probs[0], probs[1])):
#     f.write(f"{j[0]}, {j[1]}, {m[0]}, {j[1]}, {data['M4L'][n]} \n")

for i in range(NEVENT):
    f.write(f"{probs[0][0][i]}, {probs[0][1][i]}, {probs[1][0][i]}, {probs[1][1][i]}, {data['M4L'][i]} \n")

f.close()